kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt

```diff
@@ -388,12 +388,13 @@ class FunSpecTest {
 
   @Test fun expressionBodyIsDetectedReturnWithNonBreakingSpace() {
     val funSpec = FunSpec.builder("foo")
+      .returns(INT)
       .addStatement("return·1")
       .build()
 
     assertThat(funSpec.toString()).isEqualTo(
       """
-      |public fun foo() = 1
+      |public fun foo(): kotlin.Int = 1
```

Comment:
Here is an example of the behavior change.

This function previously had no return type and was relying on the fact that KotlinPoet would turn it into an expression function. If we turned that behavior off (which should be something in our power, as it shouldn't change the compiled result) the code would fail to compile.

With this behavior change, the code will now be emitted like this:
```kotlin
public fun foo(): kotlin.Unit = 1
```
which will also fail to compile. We are basically giving the expression form the same semantics as the non-expression form.

Reply:
I think this makes sense. Having to set the return type explicitly also protects from cases where you add an extra statement to the function body, which turns it from expression to block and breaks compilation.

The behaviour change is definitely something we'll want to call out in the changelog. Is it a good time to add the "unreleased" section?

Reply:
Yep, added!

Reply:
It is not supported, and any historical support was accidental and considered a bug.

Reply:
KotlinPoet doesn't support automatic type inference. There are several closed feature requests you can look up where we provide detailed justification for this design decision.

Reply:
We do not guarantee the use of an expression body as the output and so we require that you write a function body with an explicit return type and `return` statement. Whether or not the emitted code uses an expression body when producing the output is an implementation detail of the library.

Reply:
Alright, good to know. I personally would have enjoyed an option for this ‘functionality’, but I respect your decision. Thanks for the fast replies.